src/odoc/classify.ml

```diff
@@ -0,0 +1,387 @@
+(* Classify directories in ocamlfind *)
+
+(* Given a directory with cmis, cmas and so on, partition the modules between the libraries *)
+(* open Bos *)
+
+open Misc
+open Cmo_format
+
+module StringSet = Set.Make (String)
+
+let debug = ref false
+
+let log fmt =
+  if !debug then Format.printf fmt else Format.ifprintf Format.std_formatter fmt
+
+module Archive = struct
+  type name = string
+
+  type t = {
+    name : name;
+    modules : StringSet.t;
+    intf_deps : StringSet.t;
+    impl_deps : StringSet.t;
+  }
+  let empty name =
+    {
+      name;
+      modules = StringSet.empty;
+      intf_deps = StringSet.empty;
+      impl_deps = StringSet.empty;
+    }
+
+  let normalise s =
+    {
+      s with
+      intf_deps = StringSet.diff s.intf_deps s.modules;
+      impl_deps = StringSet.diff s.impl_deps s.modules;
+    }
+
+  let add_cu lib cu =
+    normalise
+      {
+        lib with
+        modules = StringSet.add cu.cu_name lib.modules;
+        intf_deps =
+          List.fold_left
+            (fun deps (cu, _) -> StringSet.add cu deps)
+            lib.intf_deps cu.cu_imports;
+        impl_deps =
+          List.fold_left
+            (fun deps id -> StringSet.add (Ident.name id) deps)
+            lib.impl_deps cu.cu_required_globals;
+      }
```

Comment:
There is some compatibility issues with OCaml `5.2` here...

The 5.2 version would be:

```suggestion
      {
        lib with
        modules =
          StringSet.add (match cu.cu_name with Compunit s -> s) lib.modules;
        intf_deps =
          List.fold_left
            (fun deps (cu, _) -> StringSet.add cu deps)
            lib.intf_deps cu.cu_imports;
        impl_deps =
          List.fold_left
            (fun deps id -> StringSet.add id deps)
            lib.impl_deps
            (List.map (fun (Compunit s) -> s) cu.cu_required_compunits);
      }
```